// @aws-quickstart/eks-blueprints/dist/addons/xray-adot-addon/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrayAdotAddOn = void 0;
const utils_1 = require("../../utils");
const adot_1 = require("../adot");
const kubectl_provider_1 = require("../helm-addon/kubectl-provider");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    deploymentMode: "deployment" /* xrayDeploymentMode.DEPLOYMENT */,
    name: 'adot-collector-xray',
    namespace: 'default'
};
/**
 * Implementation of XRAY ADOT add-on for EKS Blueprints. Installs ADOT Collector.
 */
class XrayAdotAddOn {
    constructor(props) {
        this.xrayAddOnProps = { ...defaultProps, ...props };
    }
    deploy(clusterInfo) {
        const cluster = clusterInfo.cluster;
        let doc;
        // Applying manifest for configuring ADOT Collector for Xray.
        doc = (0, utils_1.readYamlDocument)(__dirname + '/collector-config-xray.ytpl');
        const manifest = doc.split("---").map(e => (0, utils_1.loadYaml)(e));
        const values = {
            awsRegion: cluster.stack.region,
            deploymentMode: this.xrayAddOnProps.deploymentMode,
            namespace: this.xrayAddOnProps.namespace
        };
        const manifestDeployment = {
            name: this.xrayAddOnProps.name,
            namespace: this.xrayAddOnProps.namespace,
            manifest,
            values
        };
        const kubectlProvider = new kubectl_provider_1.KubectlProvider(clusterInfo);
        const statement = kubectlProvider.addManifest(manifestDeployment);
        return Promise.resolve(statement);
    }
}
__decorate([
    (0, utils_1.dependable)(adot_1.AdotCollectorAddOn.name)
], XrayAdotAddOn.prototype, "deploy", null);
exports.XrayAdotAddOn = XrayAdotAddOn;
