// @aws-quickstart/eks-blueprints/dist/addons/core-addon/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreAddOn = exports.CoreAddOnProps = void 0;
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
const utils_1 = require("../../utils");
class CoreAddOnProps {
}
exports.CoreAddOnProps = CoreAddOnProps;
const DEFAULT_NAMESPACE = "kube-system";
/**
 * Implementation of EKS Managed add-ons.
 */
class CoreAddOn {
    constructor(coreAddOnProps) {
        this.coreAddOnProps = coreAddOnProps;
    }
    deploy(clusterInfo) {
        var _a, _b;
        let serviceAccountRoleArn = undefined;
        let serviceAccount = undefined;
        let saNamespace = undefined;
        saNamespace = DEFAULT_NAMESPACE;
        if ((_a = this.coreAddOnProps) === null || _a === void 0 ? void 0 : _a.namespace) {
            saNamespace = this.coreAddOnProps.namespace;
        }
        // Create a service account if user provides namespace, PolicyDocument
        if ((_b = this.coreAddOnProps) === null || _b === void 0 ? void 0 : _b.policyDocumentProvider) {
            const policyDoc = this.coreAddOnProps.policyDocumentProvider(clusterInfo.cluster.stack.partition);
            serviceAccount = (0, utils_1.createServiceAccount)(clusterInfo.cluster, this.coreAddOnProps.saName, saNamespace, policyDoc);
            serviceAccountRoleArn = serviceAccount.role.roleArn;
        }
        let addOnProps = {
            addonName: this.coreAddOnProps.addOnName,
            addonVersion: this.coreAddOnProps.version,
            clusterName: clusterInfo.cluster.clusterName,
            serviceAccountRoleArn: serviceAccountRoleArn,
            resolveConflicts: "OVERWRITE"
        };
        const cfnAddon = new aws_eks_1.CfnAddon(clusterInfo.cluster.stack, this.coreAddOnProps.addOnName + "-addOn", addOnProps);
        if (serviceAccount) {
            cfnAddon.node.addDependency(serviceAccount);
        }
        // Instantiate the Add-on
        return Promise.resolve(cfnAddon);
    }
}
exports.CoreAddOn = CoreAddOn;
