// @aws-quickstart/eks-blueprints/dist/addons/falco/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FalcoAddOn = void 0;
const index_1 = require("../helm-addon/index");
const utils_1 = require("../../utils");
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    name: "blueprints-falco-addon",
    namespace: "falco",
    chart: "falco",
    version: "2.0.15",
    release: "falco",
    repository: "https://falcosecurity.github.io/charts",
    createNamespace: true,
    values: {}
};
/**
 * This add-on is currently not supported. It will apply the latest falco helm chart but the latest AMI does not have stock driver supported and
 * driver build in the init fails atm.
 */
class FalcoAddOn extends index_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        let values = populateValues(this.options);
        const chart = this.addHelmChart(clusterInfo, values);
        return Promise.resolve(chart);
    }
}
exports.FalcoAddOn = FalcoAddOn;
/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions) {
    var _a, _b, _c, _d, _e;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, utils_1.setPath)(values, "kubernetes.support.enabled", (_b = helmOptions.kubernetesSupportEnabled) !== null && _b !== void 0 ? _b : true);
    (0, utils_1.setPath)(values, "falco.sidekick.enabled", (_c = helmOptions.falcoSidekickEnabled) !== null && _c !== void 0 ? _c : true);
    (0, utils_1.setPath)(values, "falco.sidekick.webui.enabled", (_d = helmOptions.falcoSidekickWebuiEnabled) !== null && _d !== void 0 ? _d : true);
    (0, utils_1.setPath)(values, "audit.logs.enabled", (_e = helmOptions.auditLogsEnabled) !== null && _e !== void 0 ? _e : true);
    return values;
}
