// @aws-quickstart/eks-blueprints/dist/utils/constraints-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateConstraints = exports.ArrayConstraint = exports.NumberConstraint = exports.UrlStringConstraint = exports.StringConstraint = void 0;
const zod_1 = require("zod");
/**
 * This validates if the given string (value) is within the bounds of min to max inclusive. If not a detailed Zod Error is thrown also utilizing the identifier for context.
 */
class StringConstraint {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
    validate(key, value, identifier) {
        var _a, _b;
        if (value != undefined)
            zod_1.z.string()
                .min((_a = this.min) !== null && _a !== void 0 ? _a : 0, { message: `${key} (${identifier}: ${value}) must be no less than ${this.min} characters long.` })
                .max((_b = this.max) !== null && _b !== void 0 ? _b : 63, { message: `${key} (${identifier}: ${value}) must be no more than ${this.max} characters long.` })
                .parse(value);
    }
}
exports.StringConstraint = StringConstraint;
/**
 * This is the same as StringConstraint, but also checks if the given string is a correctly formatted URL. If not a detailed Zod Error is thrown also utilizing the identifier for context.
 */
class UrlStringConstraint {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
    validate(key, value, identifier) {
        var _a, _b;
        if (value != undefined) {
            zod_1.z.string().url({ message: `${key} (${identifier}: ${value}) must be a URL formatted correctly.` }).parse(value);
            zod_1.z.string()
                .min((_a = this.min) !== null && _a !== void 0 ? _a : 0, { message: `${key} (${identifier}: ${value}) must be a URL no less than ${this.min} characters long.` })
                .max((_b = this.max) !== null && _b !== void 0 ? _b : 100, { message: `${key} (${identifier}: ${value}) must be a URL no more than ${this.max} characters long.` })
                .parse(value);
        }
    }
}
exports.UrlStringConstraint = UrlStringConstraint;
/**
 * This class checks if the given number (value) is within the bounds of the given min and max inclusive number bounds. If not a detailed Zod Error is thrown also utilizing the identifier for context.
 */
class NumberConstraint {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
    validate(key, value, identifier) {
        var _a, _b;
        if (value != undefined)
            zod_1.z.number()
                .gte((_a = this.min) !== null && _a !== void 0 ? _a : 1, { message: `${key} (${identifier}: ${value}) must be no less than ${this.min} nodes.` })
                .lte((_b = this.max) !== null && _b !== void 0 ? _b : 3, { message: `${key} (${identifier}: ${value}) must be no more than ${this.max} nodes.` })
                .parse(value);
    }
}
exports.NumberConstraint = NumberConstraint;
/**
 * This works just like NumberConstraint but checks the length of the given value for an expected array. If not a detailed Zod Error is thrown also utilizing the identifier for context.
 */
class ArrayConstraint {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
    validate(key, value, identifier) {
        var _a, _b;
        if (value != undefined)
            zod_1.z.number()
                .gte((_a = this.min) !== null && _a !== void 0 ? _a : 1, { message: `${key} (${identifier} of length: ${value.length}) must be no less than ${this.min} node groups.` })
                .lte((_b = this.max) !== null && _b !== void 0 ? _b : 3, { message: `${key} (${identifier} of length: ${value.length}) must be no more than ${this.max} node groups.` })
                .parse(value.length);
    }
}
exports.ArrayConstraint = ArrayConstraint;
/**
 * This function validates the given object by the given constraints, and returns an error that uses the given context if needed.
 * @param constraints This is the keys of the object with specified values for validation.
 * @param context Object type name for error context purposes.
 * @param object The given object type, an array of or only a single object, to be validated.
 * @returns throws a Zod Error if validations are broken.
 */
function validateConstraints(constraints, context, ...object) {
    if (object != undefined)
        for (let i = 0; i < object.length; i++) {
            for (let k in constraints) {
                const constraint = constraints[k];
                constraint.validate(context + "." + k, object[i][k], k);
            }
        }
}
exports.validateConstraints = validateConstraints;
