// @aws-quickstart/eks-blueprints/dist/addons/calico/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalicoAddOn = void 0;
const dot = require("dot-object");
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    name: 'calico-addon',
    namespace: 'kube-system',
    version: '0.3.10',
    chart: "aws-calico",
    release: "blueprints-addon-calico",
    repository: "https://aws.github.io/eks-charts"
};
/**
 * @deprecated use CalicoOperator add-on instead
 */
class CalicoAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const values = (_a = this.options.values) !== null && _a !== void 0 ? _a : {};
        const defaultValues = {};
        dot.set("calico.node.resources.requests.memory", "64Mi", defaultValues, true);
        dot.set("calico.node.resources.limits.memory", "100Mi", defaultValues, true);
        const merged = (0, ts_deepmerge_1.default)(defaultValues, values);
        this.addHelmChart(clusterInfo, merged);
    }
}
exports.CalicoAddOn = CalicoAddOn;
