// @aws-quickstart/eks-blueprints/dist/addons/aws-privateca-issuer/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWSPrivateCAIssuerAddon = void 0;
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
const utils_1 = require("../../utils");
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    name: "blueprints-aws-pca-issuer-addon",
    chart: "aws-privateca-issuer",
    namespace: "aws-pca-issuer",
    version: "1.2.2",
    release: "aws-pca-issuer",
    repository: "https://cert-manager.github.io/aws-privateca-issuer",
    values: {},
    serviceAccountName: "aws-pca-issuer",
    iamPolicies: ["AWSCertificateManagerPrivateCAFullAccess"]
};
/**
 * Main class to instantiate the Helm chart
 */
class AWSPrivateCAIssuerAddon extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    // AWSPrivateCAIssuerAddon requires CertManagerAddOn as a prerequisite . Pls refer to documentation for more details
    deploy(clusterInfo) {
        var _a;
        //Create Service Account with IRSA
        const cluster = clusterInfo.cluster;
        let values = populateValues(this.options);
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const chart = this.addHelmChart(clusterInfo, values);
        const namespace = (0, utils_1.createNamespace)(this.options.namespace, cluster);
        if (this.options.iamPolicies.length > 0) {
            //Create Service Account with IRSA
            const opts = { name: this.options.serviceAccountName, namespace: this.options.namespace };
            const sa = cluster.addServiceAccount(this.options.serviceAccountName, opts);
            setRoles(sa, this.options.iamPolicies);
            sa.node.addDependency(namespace);
            chart.node.addDependency(sa);
        }
        return Promise.resolve(chart);
    }
}
__decorate([
    (0, utils_1.dependable)('CertManagerAddOn')
], AWSPrivateCAIssuerAddon.prototype, "deploy", null);
exports.AWSPrivateCAIssuerAddon = AWSPrivateCAIssuerAddon;
/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions) {
    var _a;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, utils_1.setPath)(values, "serviceAccount.create", helmOptions.iamPolicies.length > 0 ? false : true);
    (0, utils_1.setPath)(values, "serviceAccount.name", helmOptions.serviceAccountName);
    return values;
}
/**
 * This function will set the roles to Service Account
 * @param sa - Service Account Object
 * @param irsaRoles - Array  of Managed IAM Policies
 */
function setRoles(sa, irsaRoles) {
    irsaRoles.forEach((policyName) => {
        const policy = aws_iam_1.ManagedPolicy.fromAwsManagedPolicyName(policyName);
        sa.role.addManagedPolicy(policy);
    });
}
