// @aws-quickstart/eks-blueprints/dist/addons/calico-operator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalicoOperatorAddOn = void 0;
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    name: 'calico-operator',
    namespace: 'calico-operator',
    version: 'v3.23.3',
    chart: "tigera-operator",
    release: "bp-addon-calico-operator",
    repository: "https://projectcalico.docs.tigera.io/charts"
};
class CalicoOperatorAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const values = (_a = this.options.values) !== null && _a !== void 0 ? _a : {};
        const defaultValues = {};
        const merged = (0, ts_deepmerge_1.default)(defaultValues, values);
        this.addHelmChart(clusterInfo, merged);
    }
}
exports.CalicoOperatorAddOn = CalicoOperatorAddOn;
