// @aws-quickstart/eks-blueprints/dist/addons/container-insights/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerInsightsAddOn = void 0;
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const ts_deepmerge_1 = require("ts-deepmerge");
const __1 = require("../..");
const helm_addon_1 = require("../helm-addon");
const defaultProps = {
    name: "adot-exporter-for-eks-on-ec2",
    namespace: undefined,
    chart: "adot-exporter-for-eks-on-ec2",
    version: "0.1.0",
    release: "adot-eks-addon",
    repository: "https://aws-observability.github.io/aws-otel-helm-charts"
};
/**
 * @deprecated Use CloudWatchAdotAddOn.
 */
class ContainerInsightsAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
    }
    /**
     * @override
     */
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        const nodeGroups = (0, __1.assertEC2NodeGroup)(clusterInfo, ContainerInsightsAddOn.name);
        const policy = aws_iam_1.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy');
        nodeGroups.forEach(nodeGroup => {
            nodeGroup.role.addManagedPolicy(policy);
        });
        let values = {
            awsRegion: cluster.stack.region,
            clusterName: cluster.clusterName,
            fluentbit: {
                enabled: true
            }
        };
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const chart = this.addHelmChart(clusterInfo, values, true, false);
        return Promise.resolve(chart);
    }
}
exports.ContainerInsightsAddOn = ContainerInsightsAddOn;
